#!/usr/bin/env python3

import sys
import os
sys.path.append(os.path.dirname(os.path.abspath(__file__)))
sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

def test_grpc_connection():
    """Teste simples de conex√£o gRPC"""
    
    try:
        import grpc
        from grpc_services import services_pb2, services_pb2_grpc
        
        print("üîç Testando conex√£o gRPC...")
        
        # Testar Chatbot (porta 8082)
        try:
            print("   ü§ñ Testando Chatbot (localhost:8082)...")
            channel = grpc.insecure_channel('localhost:8082')
            stub = services_pb2_grpc.ChatbotServiceStub(channel)
            response = stub.GetStatus(services_pb2.Empty(), timeout=5)
            print(f"   ‚úÖ Chatbot: {response.message}")
            channel.close()
        except grpc.RpcError as e:
            print(f"   ‚ùå Chatbot erro: {e.code()} - {e.details()}")
        except Exception as e:
            print(f"   ‚ùå Chatbot erro geral: {e}")
        
        # Testar Cursos (porta 8081)
        try:
            print("   üìö Testando Cursos (localhost:8081)...")
            channel = grpc.insecure_channel('localhost:8081')
            stub = services_pb2_grpc.CursosServiceStub(channel)
            response = stub.GetStatus(services_pb2.Empty(), timeout=5)
            print(f"   ‚úÖ Cursos: {response.message}")
            channel.close()
        except grpc.RpcError as e:
            print(f"   ‚ùå Cursos erro: {e.code()} - {e.details()}")
        except Exception as e:
            print(f"   ‚ùå Cursos erro geral: {e}")
        
        # Testar CPAR (porta 8083)
        try:
            print("   üìÖ Testando CPAR (localhost:8083)...")
            channel = grpc.insecure_channel('localhost:8083')
            stub = services_pb2_grpc.CPARServiceStub(channel)
            response = stub.GetStatus(services_pb2.Empty(), timeout=5)
            print(f"   ‚úÖ CPAR: {response.message}")
            channel.close()
        except grpc.RpcError as e:
            print(f"   ‚ùå CPAR erro: {e.code()} - {e.details()}")
        except Exception as e:
            print(f"   ‚ùå CPAR erro geral: {e}")
            
    except ImportError as e:
        print(f"‚ùå Erro de import: {e}")
        print("   Verifique se os arquivos gRPC foram gerados corretamente")
    except Exception as e:
        print(f"‚ùå Erro geral: {e}")

def test_chatbot_functionality():
    """Teste espec√≠fico do chatbot"""
    
    try:
        import grpc
        from grpc_services import services_pb2, services_pb2_grpc
        
        print("\nü§ñ Testando funcionalidade do Chatbot...")
        
        channel = grpc.insecure_channel('localhost:8082')
        stub = services_pb2_grpc.ChatbotServiceStub(channel)
        
        # Teste ResolveDuvida
        request = services_pb2.ChatbotDuvidaRequest(
            aula_contexto="Python B√°sico",
            duvida="Como criar uma fun√ß√£o em Python?"
        )
        
        response = stub.ResolveDuvida(request, timeout=30)
        print(f"   üìù Resposta: {response.resposta}")
        
        channel.close()
        
    except Exception as e:
        print(f"   ‚ùå Erro: {e}")

if __name__ == '__main__':
    print("üß™ TESTE DE CONECTIVIDADE gRPC")
    print("=" * 40)
    
    test_grpc_connection()
    
    # Perguntar se quer testar chatbot
    resposta = input("\nü§î Testar funcionalidade do Chatbot? (s/n): ").lower()
    if resposta in ['s', 'sim', 'y', 'yes']:
        test_chatbot_functionality()
    
    print("\nüéØ PARA USAR:")
    print("   1. Mantenha os servidores rodando")
    print("   2. Use: python all_in_one.py (funciona sempre)")
    print("   3. Ou implemente sua pr√≥pria interface gRPC")